snapcraft/internal/repo/ua_manager.py

```diff
@@ -0,0 +1,62 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import contextlib
+import json
+import logging
+import subprocess
+from typing import Any, Dict, Optional
+
+logger = logging.getLogger(__name__)
+
+
+def _attach(ua_token: str) -> None:
+    """Attach UA token (ua will update package lists)."""
+    logger.info("Attaching UA token...")
+    subprocess.check_call(["sudo", "ua", "attach", ua_token])
+
+
+def _detach() -> None:
+    """Detach UA token (ua will update package lists)."""
+    logger.info("Detaching UA token...")
+    subprocess.check_call(["sudo", "ua", "detach"])
+
+
+def _is_attached() -> bool:
+    return _status()["attached"]
+
+
+def _status() -> Dict[str, Any]:
+    stdout = subprocess.check_output(["sudo", "ua", "status", "--format", "json"])
+    return json.loads(stdout)
+
+
+@contextlib.contextmanager
+def ua_manager(ua_token: Optional[str]):
+    """Attach and detach UA token as required.
+
+    Uses try/finally to ensure a detach on error."""
```

Comment:
The `"""` needs to be alone when closing a multiline string.

Reply:
There was a tool that verified things like this, right? Might be worth a future general PR incorporating it and fixing

Reply:
the `pep257` tool

Reply:
Ah, something didn't feel right when I looked at this last time.  Now I know. Will fix

Reply:
Fixed